### PR TITLE
Add Arbitrum discovery module

### PR DESCRIPTION
**Pull request by sdlyy**
## What's changed
- Added Arbitrum discovery module
- Template function for discovery configs 
- Indexers now operates per given chain

## What's left
-  [ ] Clean up repository methods
-  [ ] <strike>Clean up submodules and make configuration .env requirement only when given module is enabled</strike> - part of separate PR
-  [x] <strike>Handle discovery invalidation to minimum viable blockchain block where discovery can be safely run</strike>
-  [x] Refine Arbitrum Discovery config
-  [x] Clean up tests mess with chain ids after repo methods cleanup

### CODE REVIEW

File: packages/backend/src/config/Config.ts [hunk 1]
```diff
@@ -29,12 +32,12 @@ export interface HealthConfig {
   readonly commitSha?: string
 }
 
-export interface EthereumDiscoveryModuleConfig {
+export interface EthereumLikDiscoveryConfig {
```
Comment (michalsidzej) [inline packages/backend/src/config/Config.ts:35]:
```suggestion
export interface EthereumLikeDiscoveryConfig {
```

File: packages/backend/src/config/config.local.ts [hunk 2]
```diff
@@ -25,17 +28,30 @@ export function getLocalConfig(env: Env): Config {
...
+      },
+      arbitrum: {
+        startBlock: env.integer('START_BLOCK', 133212747),
+        clockIntervalMs: env.integer('CLOCK_INTERVAL_MS', 10 * 1000),
+        rpcUrl: env.string('ARBITRUM_RPC_URL'),
+        blockExplorerApiUrl: 'https://api.arbiscan.io/api',
+        blockExplorerApiKey: env.string('ARBISCAN_API_KEY'),
+        blockExplorerMinTimestamp: new UnixTime(
+          env.integer('ARBISCAN_MIN_TIMESTAMP', 0),
+        ),
```
Comment (michalsidzej) [inline packages/backend/src/config/config.local.ts:52]: Arbiscan is unreliable on the first arbitrum blocks so we can add this (not necessary now, but will be useful if we want to download the history)
```suggestion
        // ~ Timestamp of block number 0 on Arbitrum
        blockExplorerMinTimestamp: UnixTime.fromDate(new Date('2021-05-28T22:15:00Z')),
```

File: packages/backend/src/config/config.production.ts [hunk 3]
```diff
@@ -28,17 +31,30 @@ export function getProductionConfig(env: Env): Config {
...
+      },
+      arbitrum: {
+        startBlock: env.integer('START_BLOCK', 18127698),
+        clockIntervalMs: env.integer('CLOCK_INTERVAL_MS', 10 * 1000),
+        rpcUrl: env.string('ARBITRUM_RPC_URL'),
+        blockExplorerApiUrl: 'https://api.arbiscan.io/api',
+        blockExplorerApiKey: env.string('ARBISCAN_API_KEY'),
+        blockExplorerMinTimestamp: new UnixTime(
+          env.integer('ARBISCAN_MIN_TIMESTAMP', 0),
+        ),
```
Comment (michalsidzej) [inline packages/backend/src/config/config.production.ts:55]:
```suggestion
        // ~ Timestamp of block number 0 on Arbitrum
        blockExplorerMinTimestamp: UnixTime.fromDate(new Date('2021-05-28T22:15:00Z')),
```

File: packages/backend/src/config/discoveryConfig.ts [hunk 4]
```diff
@@ -3,104 +3,155 @@ import { DiscoveryConfig } from '@l2beat/discovery'
...
-    'Stargate Multisig': {
-      ignoreInWatchMode: ['nonce'],
+interface TemplateVariables {
+  name: string
+  chain: ChainId
+  initialAddresses: string[]
+  addresses: {
+    ultraLightNodeV2: string
+    endpoint: string
+    stargateMultisig: string
```
Comment (michalsidzej) [inline packages/backend/src/config/discoveryConfig.ts:13]: STG Multisig is not present on Arbitrum
```suggestion
```

File: packages/backend/src/config/discoveryConfig.ts [hunk 5]
```diff
@@ -3,104 +3,155 @@ import { DiscoveryConfig } from '@l2beat/discovery'
...
+  chain: ChainId.ARBITRUM,
+  initialAddresses: [
+    '0x4D73AdB72bC3DD368966edD0f0b2148401A178E2', // STG
+  ],
+  addresses: {
+    ultraLightNodeV2: '0x4D73AdB72bC3DD368966edD0f0b2148401A178E2',
+    endpoint: '0x3c2269811836af69497E5F486A85D7316753cf62',
+    stargateMultisig: '0x65bb797c2B9830d891D87288F029ed8dACc19705', // FIXME: not sure
+    layerZeroRelayer: '0x177d36dbe2271a4ddb2ad8304d82628eb921d790', // FIXME: not sure
+    layerZeroOracle: '0x5a54fe5234E811466D5366846283323c954310B2', // FIXME: not sure
```
Comment (michalsidzej) [inline packages/backend/src/config/discoveryConfig.ts:151]: We can get rid of those three (also on Ethereum)

**Review by michalsidzej — COMMENTED** [review]
Looking good, just few comments

File: packages/backend/src/config/discoveryConfig.ts (hunk 5 above)
Comment (sdlyy) [reply packages/backend/src/config/discoveryConfig.ts]: [chore: fix configuration & inteface naming](https://github.com/l2beat/lz-monitoring/pull/16/commits/87f49ca01c29da3a02163278563be390c7145f73)

File: packages/backend/src/config/config.production.ts (hunk 3 above)
Comment (sdlyy) [reply packages/backend/src/config/config.production.ts]: [chore: fix configuration & inteface naming](https://github.com/l2beat/lz-monitoring/pull/16/commits/87f49ca01c29da3a02163278563be390c7145f73)

File: packages/backend/src/config/config.local.ts (hunk 2 above)
Comment (sdlyy) [reply packages/backend/src/config/config.local.ts]: [chore: fix configuration & inteface naming](https://github.com/l2beat/lz-monitoring/pull/16/commits/87f49ca01c29da3a02163278563be390c7145f73)

File: packages/backend/src/config/Config.ts (hunk 1 above)
Comment (sdlyy) [reply packages/backend/src/config/Config.ts]: [chore: fix configuration & inteface naming](https://github.com/l2beat/lz-monitoring/pull/16/commits/87f49ca01c29da3a02163278563be390c7145f73)

File: packages/backend/.eslintrc [hunk 6]
```diff
@@ -1,3 +1,12 @@
...
-  "extends": "../../.eslintrc"
+  "extends": "../../.eslintrc",
+  "rules": {
+    "no-restricted-syntax": [
+      "error",
+      {
+        "selector": "ImportDeclaration[specifiers.length = 0]",
+        "message": "Empty imports are not allowed"
+      }
+    ]
```
Comment (michalsidzej) [inline packages/backend/.eslintrc:10]: Could this be in the repository root?

File: packages/backend/src/config/config.local.ts [hunk 7]
```diff
@@ -25,17 +26,31 @@ export function getLocalConfig(env: Env): Config {
...
+        rpcUrl: env.string('ETHEREUM_RPC_URL'),
+        blockExplorerApiUrl: 'https://api.etherscan.io/api',
+        blockExplorerApiKey: env.string('ETHERSCAN_API_KEY'),
+        blockExplorerMinTimestamp: new UnixTime(
+          env.integer('ETHERSCAN_MIN_TIMESTAMP', 0),
+        ),
+        discovery: ethereumDiscoveryConfig,
+      },
+      arbitrum: {
+        startBlock: env.integer('START_BLOCK', 133212747),
```
Comment (michalsidzej) [inline packages/backend/src/config/config.local.ts:43]:
```suggestion
        startBlock: env.integer('ARBITRUM_START_BLOCK', 133212747),
```

File: packages/backend/src/config/config.local.ts [hunk 8]
```diff
@@ -25,17 +26,31 @@ export function getLocalConfig(env: Env): Config {
...
-      etherscanApiUrl: 'https://api.etherscan.io/api',
-      etherscanApiKey: env.string('ETHERSCAN_API_KEY'),
-      etherscanMinTimestamp: new UnixTime(
-        env.integer('ETHERSCAN_MIN_TIMESTAMP', 0),
-      ),
-      discovery: discoveryConfig,
+    discovery: {
+      ethereum: {
+        // This is an arbitrary number so we do not start too far in the past.
+        startBlock: env.integer('START_BLOCK', 18127698),
```
Comment (michalsidzej) [inline packages/backend/src/config/config.local.ts:32]:
```suggestion
        startBlock: env.integer('ETHEREUM_START_BLOCK', 18127698),
```

File: packages/backend/src/config/config.production.ts [hunk 9]
```diff
@@ -28,17 +29,31 @@ export function getProductionConfig(env: Env): Config {
...
-      etherscanApiUrl: 'https://api.etherscan.io/api',
-      etherscanApiKey: env.string('ETHERSCAN_API_KEY'),
-      etherscanMinTimestamp: new UnixTime(
-        env.integer('ETHERSCAN_MIN_TIMESTAMP', 0),
-      ),
-      discovery: discoveryConfig,
+    discovery: {
+      ethereum: {
+        // This is an arbitrary number so we do not start too far in the past.
+        startBlock: env.integer('START_BLOCK', 18127698),
```
Comment (michalsidzej) [inline packages/backend/src/config/config.production.ts:35]:
```suggestion
        startBlock: env.integer('ETHEREUM_START_BLOCK', 18127698),
```

File: packages/backend/src/config/config.production.ts [hunk 10]
```diff
@@ -28,17 +29,31 @@ export function getProductionConfig(env: Env): Config {
...
+        rpcUrl: env.string('ETHEREUM_RPC_URL'),
+        blockExplorerApiUrl: 'https://api.etherscan.io/api',
+        blockExplorerApiKey: env.string('ETHERSCAN_API_KEY'),
+        blockExplorerMinTimestamp: new UnixTime(
+          env.integer('ETHERSCAN_MIN_TIMESTAMP', 0),
+        ),
+        discovery: ethereumDiscoveryConfig,
+      },
+      arbitrum: {
+        startBlock: env.integer('START_BLOCK', 134410848),
```
Comment (michalsidzej) [inline packages/backend/src/config/config.production.ts:46]:
```suggestion
        startBlock: env.integer('ARBITRUM_START_BLOCK', 134410848),
```

File: packages/backend/src/modules/ArbitrumDiscoveryModule.ts [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+  )
+
+  const blockchainClient = new BlockchainClient(provider, logger)
+  const discoveryEngine = createDiscoveryEngine(provider, config, 'arbitrum')
+
+  const clockIndexer = new ClockIndexer(
+    logger,
+    ChainId.ARBITRUM,
+    config.discovery.arbitrum.clockIntervalMs,
+  )
```
Comment (michalsidzej) [inline packages/backend/src/modules/ArbitrumDiscoveryModule.ts:37]: ClockIndexer could be shared, but it is not necessary. Let's keep it like this for now for simplicity.

**Review by michalsidzej — APPROVED** [review]
Please change the .env variables. Otherwise LGTM

File: packages/backend/.eslintrc (hunk 6 above)
Comment (sdlyy) [reply packages/backend/.eslintrc:10]: It also affects `import <module>` such as frontend's `import styles.css` giving false-positives. For backend, it works like a charm

File: packages/backend/src/modules/ArbitrumDiscoveryModule.ts (hunk 11 above)
Comment (sdlyy) [reply packages/backend/src/modules/ArbitrumDiscoveryModule.ts:37]: I'll make it right as a part of module factory

**Pull request merged**